Fix isinf for Float8 types without infinity representation

vibeswithkk opened the pull request:
## Issue Reference
Fixes #149002

## Summary
This PR fixes the `RuntimeError: "isinf" not implemented for 'Float8_e4m3fn'` error that occurs when using `torch.compile()` with `config.nan_asserts = True` on Float8 tensors.

## Problem Description
When TorchInductor's `nan_asserts` config is enabled, the generated code includes assertions like:
```python
assert not input.isinf().any().item()
```

For Float8 types that cannot represent infinity (e4m3fn, e4m3fnuz, e5m2fnuz, e8m0fnu), this would throw a RuntimeError because `isinf` was not implemented for these types.

## Solution
Modified `aten/src/ATen/native/TensorCompare.cpp` to handle Float8 types that cannot represent infinity by design:

- **Float8_e4m3fn**: "fn" means "finite, no infinity"
- **Float8_e4m3fnuz**: "fnuz" means "finite, no NaN, unsigned zero"
- **Float8_e5m2fnuz**: Same as above
- **Float8_e8m0fnu**: Exponent-only format, no infinity

For these types, `isinf()` now returns a tensor of all `False` values (since they cannot have infinite values by definition), matching the behavior for integral types.

**Float8_e5m2** is the only Float8 type that CAN represent infinity, and it continues to work correctly through the existing dispatch mechanism.

## Changes Made
### `aten/src/ATen/native/TensorCompare.cpp`
Added early return for Float8 types without infinity representation:
```cpp
// Float8 types without infinity representation return all false
if (self.scalar_type() == kFloat8_e4m3fn ||
    self.scalar_type() == kFloat8_e4m3fnuz ||
    self.scalar_type() == kFloat8_e5m2fnuz ||
    self.scalar_type() == kFloat8_e8m0fnu) {
  return at::zeros_like(self, at::kBool, at::MemoryFormat::Preserve);
}
```

### `test/test_isinf_float8.py` (New File)
Added comprehensive test coverage for Float8 isinf behavior.

## Test Plan
- Added `test/test_isinf_float8.py` with tests for all Float8 types
- Verified Float8_e5m2 (which can have infinity) still works correctly

vibeswithkk committed:
Fix isinf for Float8 types without infinity representation

Fixes #149002

Float8 types e4m3fn, e4m3fnuz, e5m2fnuz, and e8m0fnu cannot represent
infinity by design. Now isinf() returns all False for these types
instead of throwing RuntimeError.

This follows the existing pattern in isfinite() which already handles
e8m0fnu specially.

Test Plan:
- Added test/test_isinf_float8.py with comprehensive tests
- Verified Float8_e5m2 (which can have infinity) still works correctly

albanD reviewed (commented):
Code change sounds good, but revamping testing would be great!

vibeswithkk committed:
Address review: move Float8 isinf tests to test_unary_ufuncs.py

Per @albanD's review, moved tests from standalone file to test_unary_ufuncs.py
and refactored to use @dtypes decorator for cleaner parameterization.

Tests added:
- test_isinf_float8_no_infinity_types: verifies isinf returns False for
  Float8 types that cannot represent infinity (e4m3fn, e4m3fnuz, e5m2fnuz, e8m0fnu)
- test_isinf_float8_e5m2_can_have_infinity: verifies isinf correctly
  detects infinity for Float8_e5m2
- test_isinf_float8_empty_tensor: edge case for empty tensors

vibeswithkk commented:
Hi @albanD, just wanted to check if you had a chance to review the updated tests. I've moved them to test_unary_ufuncs.py and used the parametrize decorator as suggested. Let me know if there's anything else needed.
